lms/djangoapps/ccx/tests/test_utils.py

```diff
@@ -0,0 +1,562 @@
+        for state in [before, after]:
+            self.check_enrollment_state(state, True, self.user, True)
+        # mail was sent and to the right person
+        self.assertEqual(len(self.outbox), 1)
+        msg = self.outbox[0]
+        self.assertTrue(self.user.email in msg.recipients())
+
+    def test_enroll_non_user_no_email(self):
+        """register a non-user via email address but send no email
+        """
+        # ensure no emails are in the outbox now
+        self.assertEqual(len(self.outbox), 0)
+        test_email = "nobody@nowhere.com"
+        before, after = self.call_FUT(
+            student_email=test_email, email_students=False
+        )
+
+        # there should be a future membership set for this email address now
+        self.check_membership(email=test_email, future=True)
+        for state in [before, after]:
+            self.check_enrollment_state(state, False, None, False)
+        # ensure there are still no emails in the outbox now
+        self.assertEqual(len(self.outbox), 0)
+
+    def test_enroll_non_member_no_email(self):
+        """register a non-member but send no email"""
+        self.create_user()
+        # ensure no emails are in the outbox now
+        self.assertEqual(len(self.outbox), 0)
+        before, after = self.call_FUT(email_students=False)
+
+        # there should be a membership set for this email address now
+        self.check_membership(email=self.user.email)
+        self.check_enrollment_state(before, False, self.user, True)
+        self.check_enrollment_state(after, True, self.user, True)
+        # ensure there are still no emails in the outbox now
+        self.assertEqual(len(self.outbox), 0)
+
+    def test_enroll_member_no_email(self):
+        """enroll a member but send no email
+        """
+        self.register_user_in_ccx()
+        # ensure no emails are in the outbox now
+        self.assertEqual(len(self.outbox), 0)
+        before, after = self.call_FUT(email_students=False)
+
+        # there should be a membership set for this email address now
+        self.check_membership(email=self.user.email)
+        for state in [before, after]:
+            self.check_enrollment_state(state, True, self.user, True)
+        # ensure there are still no emails in the outbox now
+        self.assertEqual(len(self.outbox), 0)
+
+
+# TODO: deal with changes in behavior for auto_enroll
+class TestUnenrollEmail(ModuleStoreTestCase):
+    """Tests for the unenroll_email function from ccx.utils"""
+    def setUp(self):
+        super(TestUnenrollEmail, self).setUp()
+        # unbind the user created by the parent, so we can create our own when
+        # needed.
+        self.user = None
+        course = CourseFactory.create()
+        coach = AdminFactory.create()
+        role = CourseCcxCoachRole(course.id)
+        role.add_users(coach)
+        self.ccx = CcxFactory(course_id=course.id, coach=coach)
+        self.outbox = self.get_outbox()
+
+    def tearDown(self):
```

Comment:
There's no need for this tearDown: each test runs with a new TestUnenrollEmail instance anyway.